hail/src/main/scala/is/hail/expr/types/physical/PBinary.scala

```diff
@@ -71,56 +67,42 @@ class PBinary(override val required: Boolean) extends PType {
       }
     }
   }
-
-  override def byteSize: Long = 8
-
-  override def containsPointers: Boolean = true
 }
 
 object PBinary {
-  def apply(required: Boolean = false): PBinary = if (required) PBinaryRequired else PBinaryOptional
+  def apply(required: Boolean = false): PBinary = if(required) PBinaryRequired else PBinaryOptional
```

Comment:
Actually (in contrast to last comment on this, hidden from this view but visible under Conversations), it turns out that we need to do this explicit constructor.

It turns out that calling `PCanonicalBinary.apply(required)`, even though this returns a PCanonicalBinaryRequired or PCanonicalBinaryOptional, `isOfType` evaluates to false, even though `this == other` evaluates to true!! This feels like a bug, but I choose to leave the implementation alone

Reply:
Meaning, in codeOrdering:

`println(s"Other PTYPE is ${other}, and this type is ${this} and equals? ${this == other}, this comparisons: ${this == PBinaryRequired} ${this == PCanonicalBinaryRequired}, other: ${this == PCanonicalBinaryRequired} ${other == PBinaryRequired} ")")`

Prints: 
"Other PTYPE is +Binary, and this type is +Binary and equals? true, this comparisons: false true, other: true false"
  * `this == other` is `true`
  * `this == PCanonicalBinaryRequired` is `true`,  `other == PCanonicalBinaryRequired` is `false` and vice versa

For: 

PBinary:
```scala
case object PBinaryOptional extends PCanonicalBinary(false)
case object PBinaryRequired extends PCanonicalBinary(true)

# ...
object PBinary {
  def apply(required: Boolean = false): PBinary = PCanonicalBinary.apply(required)
```

PCanonicalBinary:
```scala
case object PCanonicalBinaryOptional extends PCanonicalBinary(false)
case object PCanonicalBinaryRequired extends PCanonicalBinary(true)

# ...
object PCanonicalBinary {
  def apply(required: Boolean = false): PBinary = if (required) PCanonicalBinaryRequired else PCanonicalBinaryOptional
```

Reply:
This behavior derives from how case classes and case objects define equality, I think.

Reply:
this should just call `PCanonicalBinary(required)` which dispatches to the case objects.

Reply:
Yeah what’s weird if I do that (dispatch to the case objects) equality fails as above. Your proposal should be equivalent to PCanonicalBinary.apply(required), which didn’t work. 

This is all strange because case objects are evaluated by value, not reference, https://docs.scala-lang.org/tutorials/tour/case-classes.html. But I can’t see the bug

Reply:
Which means that something isn't equal between these PBinaryOptional and PBinaryRequired.

I've used your suggestion, but this won't solve the question of what is different :(